src/frontend/src/observer/observer_manager.rs

```diff
@@ -468,6 +512,30 @@ impl FrontendObserverNode {
         self.hummock_snapshot_manager.update(deltas);
     }
 
+    fn handle_cursor_notify(&self, table_ids: Vec<TableId>) {
+        if table_ids.is_empty() {
+            return;
+        }
+        for session in self.sessions_map.read().values() {
```

Comment:
Currently one CursorNotifier per session and within each CursorNotifier, we have one channel per table. Assume we have N sessions and M tables in total, we can end up creating N*M channels. Not to mention that we have a task in CursorNotifier to maintain the per table channel. These seem inefficient and complicated. IMO, a better way to implement the notification is:
- Create one and only one watch channel to listen on table changed log updates.
- When cursor needs to wait for changelog epoch changes for a table, it only needs to subscribe to this channel.

Implementation-wise, maybe we can store HummockSnapshotManager in each cursor instance and implement a wait_for_epoch similar to the [method](https://github.com/risingwavelabs/risingwave/blob/c47a2e8fce4a1a5bbd4f422b816be2b18460215f/src/frontend/src/scheduler/snapshot.rs#L257) here.

Reply:
Actually, I think the current cursor implementation in frontend can be greatly simplified. The most complicated part in the current cursor is the state machine. The async feature of rust is a very powerful tool to generate state machine. So for a cursor, it can be implemented as a stream
```
#[try_stream(...)]
async fn cursor_stream(...) {
    if need_snapshot {
         for row in create_snapshot_iter(table_id) { yield row; }
    }
    let mut current_epoch = ...;
    loop {
        for row in change_log(current_epoch, table_id) { yield row; }
        current_epoch = snapshot_manager.wait_next_epoch(current_epoch, table_id);
    }
}
```

When a cursor is created, we can create such stream and hold it in the session. On a `fetch`, we just call `cursor_stream.next().await`, and the timeout can be applied with simply `timeout(timeout_duration, cursor_stream.next()).await`.

We can be aware of table drop in `snapshot_manager`, when we see that the `table_id` does not exist in the `state_table_info` of `FrontendHummockVersion` anymore, and then changes on catalog in this PR can be totally avoided.